// popyt/out/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CAPTION_DATA = exports.CHANNEL_SECTION_DATA = exports.CHANNEL_DATA = exports.WATERMARK_DATA = exports.COMMENT_DATA = exports.PLAYLIST_ITEM_DATA = exports.PLAYLIST_DATA = exports.SUBSCRIPTION_DATA = exports.COMMENT_THREAD_DATA = void 0;
/**
 * @ignore
 */
exports.COMMENT_THREAD_DATA = {
    snippet: {
        channelId: undefined,
        topLevelComment: {
            snippet: {
                textOriginal: undefined
            }
        }
    }
};
/**
 * @ignore
 */
exports.SUBSCRIPTION_DATA = {
    snippet: {
        resourceId: {
            kind: 'youtube#channel',
            channelId: undefined
        }
    }
};
/**
 * @ignore
 */
exports.PLAYLIST_DATA = {
    snippet: {
        title: undefined
    }
};
/**
 * @ignore
 */
exports.PLAYLIST_ITEM_DATA = {
    snippet: {
        playlistId: undefined,
        resourceId: {
            kind: 'youtube#video',
            videoId: undefined
        }
    }
};
/**
 * @ignore
 */
exports.COMMENT_DATA = {
    snippet: {
        textOriginal: undefined
    }
};
/**
 * @ignore
 */
exports.WATERMARK_DATA = {
    timing: {
        type: undefined,
        offsetMs: undefined,
        durationMs: undefined
    }
};
/**
 * @ignore
 */
exports.CHANNEL_DATA = {
    id: undefined
};
/**
 * @ignore
 */
exports.CHANNEL_SECTION_DATA = {
    snippet: {
        type: undefined,
        style: undefined
    }
};
/**
 * @ignore
 */
exports.CAPTION_DATA = {};
