// popyt/out/entities/language.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Language = void 0;
/**
 * A language.
 */
class Language {
    constructor(youtube, data) {
        this.youtube = youtube;
        this.data = data;
        this._init(data);
    }
    /**
     * @ignore
     */
    _init(data) {
        if (data.kind !== 'youtube#i18nLanguage') {
            throw new Error(`Invalid language type: ${data.kind}`);
        }
        const language = data;
        this.id = language.id;
        this.full = true;
        /* istanbul ignore next */
        if (language.snippet) {
            this.hl = language.snippet.hl;
            this.name = language.snippet.name;
        }
        else {
            /* istanbul ignore next */
            this.full = false;
        }
    }
}
exports.Language = Language;
/**
 * The name of the endpoint used for this entity.
 */
Language.endpoint = 'i18nLanguages';
/**
 * The parts to request for this entity.
 */
Language.part = 'snippet';
/**
 * The fields to request for this entity.
 */
Language.fields = 'items(kind,id,snippet(hl,name))';
