// popyt/out/entities/region.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Region = void 0;
/**
 * A region.
 */
class Region {
    constructor(youtube, data) {
        this.youtube = youtube;
        this.data = data;
        this._init(data);
    }
    /**
     * @ignore
     */
    _init(data) {
        if (data.kind !== 'youtube#i18nRegion') {
            throw new Error(`Invalid region type: ${data.kind}`);
        }
        const region = data;
        this.id = region.id;
        this.full = true;
        /* istanbul ignore next */
        if (region.snippet) {
            this.gl = region.snippet.gl;
            this.name = region.snippet.name;
        }
        else {
            /* istanbul ignore next */
            this.full = false;
        }
    }
}
exports.Region = Region;
/**
 * The name of the endpoint used for this entity.
 */
Region.endpoint = 'i18nRegions';
/**
 * The parts to request for this entity.
 */
Region.part = 'snippet';
/**
 * The fields to request for this entity.
 */
Region.fields = 'items(kind,id,snippet(gl,name))';
