// popyt/out/entities/video-abuse-report-reason.js
"use strict";
/* istanbul ignore file */
/* OAuth */
Object.defineProperty(exports, "__esModule", { value: true });
exports.VideoAbuseReportReason = void 0;
/**
 * A YouTube [[Video]] abuse report reason.
 */
class VideoAbuseReportReason {
    constructor(youtube, data) {
        this.youtube = youtube;
        this.data = data;
        this._init(data);
    }
    /**
     * @ignore
     */
    _init(data) {
        if (data.kind !== 'youtube#videoAbuseReportReason') {
            throw new Error(`Invalid video abuse report reason type: ${data.kind}`);
        }
        const reason = data;
        this.id = reason.id;
        this.full = true;
        /* istanbul ignore next */
        if (reason.snippet) {
            this.label = reason.snippet.label;
            this.secondaryReasons = reason.snippet.secondaryReasons;
        }
        else {
            /* istanbul ignore next */
            this.full = false;
        }
    }
}
exports.VideoAbuseReportReason = VideoAbuseReportReason;
/**
 * The name of the endpoint used for this entity.
 */
VideoAbuseReportReason.endpoint = 'videoAbuseReportReasons';
/**
 * The parts to request for this entity.
 */
VideoAbuseReportReason.part = 'id,snippet';
/**
 * The fields to request for this entity.
 */
VideoAbuseReportReason.fields = 'items(kind,id,snippet(label,secondaryReasons))';
